geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/ZAddExecutor.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.redis.internal.executor.sortedset;
+
+
+import static org.apache.geode.redis.internal.RedisConstants.ERROR_INVALID_ZADD_OPTION_NX_XX;
+import static org.apache.geode.redis.internal.RedisConstants.ERROR_NOT_A_VALID_FLOAT;
+import static org.apache.geode.redis.internal.RedisConstants.ERROR_SYNTAX;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.apache.geode.redis.internal.executor.AbstractExecutor;
+import org.apache.geode.redis.internal.executor.RedisResponse;
+import org.apache.geode.redis.internal.netty.Coder;
+import org.apache.geode.redis.internal.netty.Command;
+import org.apache.geode.redis.internal.netty.ExecutionHandlerContext;
+
+public class ZAddExecutor extends AbstractExecutor {
+  private final ZAddExecutorState zAddExecutorState = new ZAddExecutorState();
+
+  @Override
+  public RedisResponse executeCommand(Command command, ExecutionHandlerContext context) {
+    zAddExecutorState.initialize();
+    RedisSortedSetCommands redisSortedSetCommands = context.getRedisSortedSetCommands();
+    List<byte[]> commandElements = command.getProcessedCommand();
+    Iterator<byte[]> commandIterator = commandElements.iterator();
+
+    skipCommandAndKey(commandIterator);
+
+    byte[] firstScore = findAndValidateZAddOptions(command, commandIterator, zAddExecutorState);
+    if (zAddExecutorState.exceptionMessage != null) {
+      return RedisResponse.error(zAddExecutorState.exceptionMessage);
+    }
+
+    List<byte[]> scoresAndMembersToAdd = getScoresAndMembers(firstScore, commandIterator,
+        zAddExecutorState);
+    if (zAddExecutorState.exceptionMessage != null) {
+      return RedisResponse.error(zAddExecutorState.exceptionMessage);
+    }
+
+    return RedisResponse
+        .integer(redisSortedSetCommands.zadd(command.getKey(), scoresAndMembersToAdd,
+            makeOptions(zAddExecutorState)));
+  }
```

Comment:
Thinking about this more, we could actually avoid the extra allocation to a list here and the additional method iterating through the arguments by doing the following:

```
  return RedisResponse
        .integer(redisSortedSetCommands.zadd(command.getKey(), commandElements.subList(zAddExecutorState.optionsFoundCount + 2, commandElements.size()),
            makeOptions(zAddExecutorState)));
```

Then, in the `memberAdd` method of `RedisSortedSet`, after getting the score we could attempt to parse the double and throw an error if we are unable to:
```
try {
        Double.valueOf(Coder.bytesToString(scoreToAdd));
      } catch (NumberFormatException e) {
        throw new NumberFormatException(ERROR_NOT_A_VALID_FLOAT);
      }
```

This would also allow us to get rid of the `getScoresAndMembers` method here as well as that extra if block (lines 50-52).  We'd also no longer need to return anything from `findAndValidateZAddOptions`.